### ISSUE TITLE
Reuse target connection: failure to properly re-open connection

### ISSUE THREAD

**Issue by mistressofjellyfish** [jtpereyda/boofuzz#390]
After looking at the source code, I am quite convinced that reuse_target_connection is handled improperly when no restart_callback is passed to the Session object.

In particular, the connection gets opened in `_main_fuzzing_loop`, but if a restart occurs either via vmcontrol or a Monitor, these code paths do not contain a method to re-open  the connection after the restart. It is my understanding that in this case, the connection is dropped and needs to be reopened (if it's TCP at least).

This is also true without my work in #385.

Additionally, if `on_failure` callbacks are present, the restart_target callbacks seem not to be called for restart (same goes for the monitors). Is this by design? Why is `on_failure` even called in `_restart_target`?

**Comment by SR4ven** [comment]
I too have noticed that the whole restart code isn't very consistent. And I think I have mentioned something similar in #363 before.

I doubt that ignoring restart_target callbacks in case of on_failure presence is intended. The best option might be to run through all restart callbacks until one returns a value that indicates a successful restart.
About on_failure and EventHook, I have never used those. @jtpereyda can you help out?

**Comment by jtpereyda** [comment]
I agree with @mistressofjellyfish on both counts
1. The reuse_target_connection behavior sounds like a bug. I forget who added this and when exactly, but I've never used it myself in a practical situation.
2. I agree that the on_failure callback should be called in addition to the restart callbacks, not instead of.

The behavior of the restart methods is defined in `Session._restart_target()` -- I think this bug could be resolved by changing the first `elif` to an `if`. 

As for making the code call multiple restart methods, I could go either way. I'm not sure what use case would encourage multiple restart callbacks, but I guess calling all of them in such a situation is reasonable.

**Comment by mistressofjellyfish** [comment]
My current plan regarding the restart methods is to try them in a row until one of them reports "restart successful".

Most callbacks are integrated in the monitor infrastructure with #385, and this implies that the rules that apply for monitors apply for these callbacks as well. Theoretically, this then can do both: either provide one way to restart a target or just make the last method in the chain return true.

Should I fix this bug directly in my mega-PR #385 or should we fix this in a separate PR after #385 is merged?

**Comment by jtpereyda** [comment]
I recommend separate.

**Comment by SR4ven** [comment]
From https://github.com/jtpereyda/boofuzz/pull/385#discussion_r385598441:
Rework log messages related to opening and closing the target connection and take `_reuse_target_connection` into account.
